packages/itwin/grouping-mapping-widget/src/widget/hooks/useVisualization.ts

```diff
@@ -0,0 +1,90 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+import type { IModelConnection } from "@itwin/core-frontend";
+import { toaster } from "@itwin/itwinui-react";
+import type { ISelectionProvider, SelectionChangeEventArgs } from "@itwin/presentation-frontend";
+import { Presentation } from "@itwin/presentation-frontend";
+import { useCallback, useEffect, useState } from "react";
+import { useGroupHilitedElementsContext } from "../components/context/GroupHilitedElementsContext";
+import { visualizeGroupColors } from "../components/groupsHelpers";
+import { clearEmphasizedElements, clearOverriddenElements, transparentOverriddenElements, visualizeElementsByQuery, zoomToElements } from "../components/viewerUtils";
+
+export const useVisualization = (shouldVisualize: boolean, iModelConnection: IModelConnection, query: string, queryGenerationType: string) => {
+  const [isRendering, setIsRendering] = useState<boolean>(false);
+  const { showGroupColor, groups, hiddenGroupsIds, hilitedElementsQueryCache } = useGroupHilitedElementsContext();
+  const [simpleSelectionQuery, setSimpleSelectionQuery] = useState<string>("");
+
+  const resetView = useCallback(async () => {
+    if (!shouldVisualize) return;
+    if (showGroupColor) {
+      await visualizeGroupColors(iModelConnection, groups, hiddenGroupsIds, hilitedElementsQueryCache);
+    } else {
+      clearOverriddenElements();
+    }
+    clearEmphasizedElements();
+  }, [groups, hiddenGroupsIds, hilitedElementsQueryCache, iModelConnection, showGroupColor, shouldVisualize]);
+
+  useEffect(() => {
+    if (!shouldVisualize) return;
```

Comment:
Probably the listener should be removed if `shouldVisualize` toggles from `true` to `false`.

Reply:
It does get removed, the cleanup function of the previous effect will run before the new effect. https://react.dev/reference/react/useEffect#parameters